File: go/mongobenchmark.go
package main

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type Book struct {
	Title  string `bson:"title"`
	Author string `bson:"author"`
}

func main() {

	start := time.Now()

	client, _ := mongo.Connect(context.Background(), options.Client().ApplyURI("mongodb://localhost:27017"))
	collection := client.Database("test").Collection("go")

	// Drop collection to start from scratch
	collection.Drop(context.Background())

	// Insert documents using 8 goroutines
	var wg sync.WaitGroup

	for i := 0; i < 8; i++ {
		wg.Add(1)
		go insertMany(&wg)
	}

	wg.Wait()

	elapsed := time.Since(start)
	fmt.Printf("Insert documents took %s", elapsed)
	fmt.Println()

	// Now, find documents
	start = time.Now()

	batch_size := int32(1000)
	cursor, err := collection.Find(context.Background(), bson.M{"author": "Isaac Asimov"}, &options.FindOptions{BatchSize: &batch_size})
	if err != nil {
		log.Fatal(err)
	}

	defer cursor.Close(context.Background())
	for cursor.Next(context.Background()) {
		var document bson.M
		if err = cursor.Decode(&document); err != nil {
			log.Fatal(err)
		}

		var _ = document["title"]
		var _ = document["author"]
	}

	elapsed = time.Since(start)
	fmt.Printf("Find documents took %s", elapsed)

}

func insertMany(wg *sync.WaitGroup) {

	defer wg.Done()

	client, err := mongo.Connect(context.Background(), options.Client().ApplyURI("mongodb://localhost:27017"))

	defer func() {
		if err = client.Disconnect(context.Background()); err != nil {
			panic(err)
		}
	}()

	collection := client.Database("test").Collection("go")

	dune := Book{
		Title:  "Dune",
		Author: "Frank Herbert",
	}
	i_robot := Book{
		Title:  "I, Robot",
		Author: "Isaac Asimov",
	}
	foundation := Book{
		Title:  "Foundation",
		Author: "Isaac Asimov",
	}
	brave_new_world := Book{
		Title:  "Brave New World",
		Author: "Aldous Huxley",
	}

	books := []interface{}{dune}

	for j := 0; j < 33; j++ {
		books = append(books, i_robot, foundation, brave_new_world)
	}

	for i := 0; i < 6250; i++ {

		_, err := collection.InsertMany(context.Background(), books)

		if err != nil {
			log.Fatal(err)
		}

	}

}

File: node/mongobenchmark.js
const { MongoClient } = require("mongodb");
// Connection URI
const uri = "mongodb://localhost:27017";
// Create a new MongoClient
const client = new MongoClient(uri);

async function run() {
  try {
    var start = new Date();

    // Connect the client to the server
    await client.connect();
    // Clean collection first. If doesn't exist yet, catch the exception
    await client.db("test").collection("node").drop().catch(e => {})

    let collection = client.db("test").collection("node");

    // Insert all documents, doing batches of 8 concurrent async calls to the same client 
    for (i=0; i<6250; i++) {

      let insert_promises = [];
      
      for(t=0; t<8; t++) {
        // Doc list needs to be created each time so they are not considered the same docs by MongoClient
        // (otherwise they will be rejected by the driver due to duplication)
        let docs = [
          { "title": "Dune", "author": "Frank Herbert" },
        ];
      
        for (d=0; d<33; d++) {
          docs.push({ "title": "I, Robot", "author": "Isaac Asimov" });
          docs.push({ "title": "Foundation", "author": "Isaac Asimov" });
          docs.push({ "title": "Brave New World", "author": "Aldous Huxley" });
        }

        insert_promises.push(collection.insertMany(docs));
      }

      await Promise.all(insert_promises);
    }

    let end = new Date() - start
    console.info('Insert execution time: %dms', end)
    
    start = new Date();
    // Now let's find some of all inserted docs
    const findCursor = await collection.find({ author: "Isaac Asimov" }).batchSize(1000);

    let totalChars = 0; 
    await findCursor.forEach((doc) => {
      let author = doc.author;
      let title = doc.title;
      totalChars += author.length + title.length;
    });

    console.info('Total Chars: %d', totalChars)

    end = new Date() - start
    console.info('Fetch execution time: %dms', end)

  } finally {
    // Ensures that the client will close when you finish/error
    await client.close();
  }
}
run().catch(console.dir);